docs/ref/templates/builtins.txt

```diff
@@ -1641,9 +1641,10 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off::
 ``escapejs``
 ------------
 
-Escapes characters for use in JavaScript strings. This does *not* make the
-string safe for use in HTML or JavaScript template literals, but does protect
-you from syntax errors when using templates to generate JavaScript/JSON.
+Escapes characters for use in JavaScript strings. This makes the string safe
+for use in HTML and JavaScript or JSON string literals. Note that it does *not*
+make the string safe for use in "JavaScript template literals" (i.e. the
+JavaScript backtick syntax).
```

Comment:
I don't think this is right. 🤔

ticket-27900 made escapejs escape backticks to support usage in ES6 template literals. 6ae1b04fb584db0fdb22b8e287784c4ed3ac62ac

Also `safe` may not be right... i.e. is that not what's causing the confusion here? It means you won't get a syntax error using the DTL to interpolate, but it doesn't protect against XSS &co, which is what I guess _safe_ might be taken for. 

Maybe deprecating `escapejs`... — I suspect folks might complain 🤔

Reply:
> I don't think this is right. 🤔
> 
> [ticket-27900](https://code.djangoproject.com/ticket/27900) made escapejs escape backticks to support usage in ES6 template literals. https://github.com/django/django/commit/6ae1b04fb584db0fdb22b8e287784c4ed3ac62ac

Eh? It was you yourself who pointed me to ticket-29055 which explains how escapejs does not work with ES6 template literals.

> Also safe may not be right... i.e. is that not what's causing the confusion here? It means you won't get a syntax error using the DTL to interpolate, but it doesn't protect against XSS &co, which is what I guess safe might be taken for.

Yes that is what "safe" means and how I am using it. Either the filter makes the content "safe" or it is useless for almost any purpose. escapejs does protect against XSS.

> Maybe deprecating escapejs... — I suspect folks might complain 🤔

I'm trying to do the opposite - to rejuvenate this useful feature that has languished unnecessarily due to its documentation being incomprehensible / incorrect.

Reply:
> Yes that is what "safe" means and how I am using it. 

OK, but if that's the meaning of "safe" then there's no usage of `escapejs` which is _safe_. It simply doesn't do that. AFAICS the only use of `escapejs` is to allow you to interpolate values from DTL template variables without needing to worry about JS syntax errors, when they contain `'` for example. 

It's strikes me the existing is quite clear on this: 

> This does *not* make the string safe for use in HTML or JavaScript template literals, but does protect
 you from syntax errors when using templates to generate JavaScript/JSON.

I kind of (still) think we should close this as `wontfix`. 

This is a difficult problem. The `json_script` tag is the way to go. https://docs.djangoproject.com/en/4.0/ref/templates/builtins/

Perhaps @adamchainz can comment, having blogged on this topic? 🙏

Reply:
I did indeed only look at as an afterthought. I don't write inline `<script>` tags ever, because they are inefficient and disallowed with a strong CSP.

The way I understood `escapejs` as being safe is for templated whole javascript files, i.e. where the template is the JS file. This may indeed be a mismatch between the docs and what the code does, as the code does escape `<` & co., so it looks like it could be safe for templating inline `<script>`s.

I'm not sure what the previous authors of the docs meant by "not safe for use in HTML". Perhaps there are cases I haven't thought of. It would be better to contact them to help with clarification.

Reply:
`escapejs` was introduced 14 years ago for ticket-4131. 

The [original diff](https://github.com/django/django/commit/b65fce659502ac5e211d13fbd5435e1ff6c703d2#diff-d247cb923bcb90ec15cafc2d94d1ba7a1bf7da7a89175902adc7431f5475a9dcR1312-R1314) has that it is **not** _safe_ for use in HTML, just to avoid syntax errors. 

> This does *not* make the string safe for use in HTML, but does protect you from syntax errors when using
 templates to generate JavaScript/JSON.

The ticket discusses that it doesn't protect correctly against XSS — though **some** protection is granted — hence the disclaimer.

Reply:
Well I think the mystery is solved then. In the original diff you linked where escapejs was introduced, the way it worked was completely different from the way it works now, and probably was indeed not safe for HTML.

But in ticket-7177 the code was updated to change the way it works, the idea being to make it so that it _is_ safe for HTML. However the patch that implemented this, 8c4a525871df19163d5bfdf5939eff33b544c2e2, forgot to update the documentation.

So the situation is: the documentation was written for the original version of escapejs, but since then multiple changes have been made to the code which have caused the documentation and the implementation to diverge. So my PR is simply updating the documentation to reflect the code as it exists now.

Reply:
OK, I'll ask the security team to look.

Reply:
Hello everyone! I was trying to fully understand this conversation to try to provide a proposal to unblock this. I was about to reach out to the security team, following the latest comment from @carltongibson, but before doing so I tried to find a way to produce **unsafe** content even when using `escapejs`. I think I found a way, and that this PR should be closed without merging. My example is as follows:

```python
def test_escapejs(self):
    from django.template import Template, Context

    t = Template(
        "{{ var1|escapejs }}\n"
        "<script>{{ var2|escapejs }}</script>\n"
        "<script>{{ var3|escapejs }}</script>\n"
    )
    c = Context({
        "var1": "<script>alert('XSS 1');</script>",
        "var2": 'alert("XSS 2");',
        "var3": "alert(String.fromCharCode(88,83,83))",
    })
    response = t.render(c)
    return HttpResponse(response)
```

The code above would generate:

```html
\u003Cscript\u003Ealert(\u0027XSS 1\u0027)\u003B\u003C/script\u003E
<script>alert(\u0022XSS 2\u0022)\u003B</script>
<script>alert(String.fromCharCode(88,83,83))</script>
```
and the third line is indeed **unsafe**, since the alert is executed. So my advice is to close this PR and do not change the documentation (if anything, discourage the `escapejs` usage or rename it to something that implies "less safeness".